### PR TITLE
rgw/pubsub: notification filtering by object tags

### PR DESCRIPTION
**Pull request by yuvalif**
Signed-off-by: Yuval Lifshitz <yuvalif@yahoo.com>

* This PR adds object tags info to notifications on object creation events.
* This PR also adds the ability to filter notifications based on the objects tags.

> This also fixes an existing issue with metadata info not appearing in deletion requests. Tracker Issue: https://tracker.ceph.com/issues/43196

please see the updated docs for more info on the feature.

### CODE REVIEW

File: doc/radosgw/notifications.rst [hunk 1]
```diff
@@ -298,6 +299,7 @@ pushed or pulled using the pubsub sync module.
 - s3.object.version: object version in case of versioned bucket
 - s3.object.sequencer: monotonically increasing identifier of the change per object (hexadecimal format)
 - s3.object.metadata: any metadata set on the object sent as: ``x-amz-meta-`` (an extension to the S3 notification API) 
+- s3.object.tags: any tags set on the objet. Will be sent only in object creation events (an extension to the S3 notification API)
```
Comment (ofriedma) [inline doc/radosgw/notifications.rst]: object

File: doc/radosgw/notifications.rst [hunk 2]
```diff
@@ -273,7 +273,8 @@ pushed or pulled using the pubsub sync module.
                    "eTag":"",
                    "versionId":"",
                    "sequencer": "",
-                   "metadata":""
+                   "metadata":"",
```
Comment (ofriedma) [inline doc/radosgw/notifications.rst]: I think that for metadata and tags we should add exactly how to add the Key values in the XML

File: doc/radosgw/pubsub-module.rst [hunk 3]
```diff
@@ -474,7 +475,8 @@ the events will have an S3-compatible record format (JSON):
                    "eTag":"",
                    "versionId":"",
                    "sequencer":"",
-                   "metadata":""
+                   "metadata":"",
```
Comment (ofriedma) [inline doc/radosgw/pubsub-module.rst]: We should claim that Metadata or tags are not supported for deletion notifications and I think we should check that any new XML for notifications delete events, should not include Metadata or tags.

File: doc/radosgw/notifications.rst (hunk 2 above)
Comment (yuvalif) [reply doc/radosgw/notifications.rst]: the xml structure is documented: https://docs.ceph.com/docs/master/radosgw/s3/bucketops/#create-notification
but you can use boto3 or aws CLI by adding the updated extensions file.
instructions are here: https://github.com/ceph/ceph/blob/70047a93e3d70eeb6a5db7a291cd70a9f5cec1a2/examples/boto3/README.md

File: doc/radosgw/pubsub-module.rst (hunk 3 above)
Comment (yuvalif) [reply doc/radosgw/pubsub-module.rst]: agree, will ad a check to reject notifications with this combinations.
this applied to tags and metadata in the same way.
will update documentation around metadata as well.

Comment (yuvalif) [reply doc/radosgw/pubsub-module.rst]: @ofriedma both metadata and tags are now supported in deletion notifications

**Comment by yuvalif** [comment]
all teuthology runs are failing: http://pulpito.front.sepia.ceph.com/yuvalif-2019-11-28_08:09:16-rgw:verify-wip-yuval-tag-filter-distro-basic-smithi/
however, same failures exist when running teuthology against master: http://pulpito.front.sepia.ceph.com/yuvalif-2019-11-29_09:46:32-rgw:verify-wip-yuval-master-distro-basic-smithi/

File: src/rgw/rgw_pubsub.h [hunk 4]
```diff
@@ -109,6 +115,13 @@ class rgw_pubsub_topic_filter;
           <Value></Value>
         </FilterRule>
       </S3Metadata>
+      <S3ObjectTags>
+        <FilterRule>
+          <Name></Name>
+          <Value></Value>
+        </FilterRule>
+      </s3ObjectTags>
+>>>>>>> rgw/pubsub: notification filtering by object tags
```
Comment (cbodley) [inline src/rgw/rgw_pubsub.h]: git happened here

Comment (yuvalif) [reply src/rgw/rgw_pubsub.h]: > git happened here

fixed

**Comment by yuvalif** [comment]
jenkins test make check arm64

**Comment by yuvalif** [comment]
@cbodley any other issues with this PR?

**Pull request merged**